variable/rebin.cpp

```diff
@@ -82,29 +101,49 @@ Variable rebin(const VariableConstView &var, const Dim dim,
     throw except::BinEdgeError(
         "The input does not have coordinates with bin-edges.");
 
-  if (var.dims().inner() == dim) {
-    using namespace rebin_inner_detail;
-    return transform_subspan<std::tuple<
-        args<double, double, double, double>, args<float, float, float, float>,
-        args<float, double, float, double>, args<float, float, float, double>,
-        args<bool, double, bool, double>>>(var.dtype(), dim,
-                                           newCoord.dims()[dim] - 1, newCoord,
-                                           var, oldCoord, core::element::rebin);
+  using transform_args = std::tuple<
+      args<double, double, double, double>, args<float, float, float, float>,
+      args<float, double, float, double>, args<float, float, float, double>,
+      args<bool, double, bool, double>>;
 
+  const bool ascending = is_sorted(oldCoord, dim, SortOrder::Ascending) &&
+                         is_sorted(newCoord, dim, SortOrder::Ascending);
+  if (!ascending && !(is_sorted(oldCoord, dim, SortOrder::Descending) &&
+                      is_sorted(newCoord, dim, SortOrder::Descending)))
+    throw except::BinEdgeError(
+        "Rebin: The old or new bin edges are not sorted.");
+  if (var.dims().inner() == dim) {
+    if (ascending) {
+      return transform_subspan<transform_args>(
+          var.dtype(), dim, newCoord.dims()[dim] - 1, newCoord, var, oldCoord,
+          core::element::rebin<Less>);
+    } else {
+      return transform_subspan<transform_args>(
+          var.dtype(), dim, newCoord.dims()[dim] - 1, newCoord, var, oldCoord,
+          core::element::rebin<Greater>);
+    }
   } else {
     auto dims = var.dims();
     dims.resize(dim, newCoord.dims()[dim] - 1);
     Variable rebinned(var, dims);
     if (newCoord.dims().ndim() > 1)
       throw std::runtime_error(
           "Not inner rebin works only for 1d coordinates for now.");
-    if (oldCoord.dtype() == dtype<double>)
-      rebin_non_inner<double>(dim, var, rebinned, oldCoord, newCoord);
-    else if (oldCoord.dtype() == dtype<float>)
-      rebin_non_inner<float>(dim, var, rebinned, oldCoord, newCoord);
-    else
+    if (oldCoord.dtype() == dtype<double>) {
+      if (ascending)
+        rebin_non_inner<double, Less>(dim, var, rebinned, oldCoord, newCoord);
+      else
+        rebin_non_inner<double, Greater>(dim, var, rebinned, oldCoord,
+                                         newCoord);
+    } else if (oldCoord.dtype() == dtype<float>) {
+      if (ascending)
+        rebin_non_inner<float, Less>(dim, var, rebinned, oldCoord, newCoord);
+      else
+        rebin_non_inner<float, Greater>(dim, var, rebinned, oldCoord, newCoord);
```

Comment:
Minor point. Very similar to condition above could consider implementation with type template argument.

Reply:
Not sure what you suggest? We need to transition from a runtime argument to a compile-time branch somehow. Which solution do you have in mind?

Reply:
Indeed. I was thinking of a replacement more along these lines.

```
    if (oldCoord.dtype() == dtype<double>) {
      if (ascending)
        rebin_non_inner<double, Less>(dim, var, rebinned, oldCoord, newCoord);
      else
        rebin_non_inner<double, Greater>(dim, var, rebinned, oldCoord,
                                         newCoord);
    } else if (oldCoord.dtype() == dtype<float>) {
      if (ascending)
        rebin_non_inner<float, Less>(dim, var, rebinned, oldCoord, newCoord);
      else
        rebin_non_inner<float, Greater>(dim, var, rebinned, oldCoord, newCoord);
    } else {
      throw std::runtime_error(
          "Rebinning is possible only for double and float types.");
    }
```

to
```
    if (oldCoord.dtype() == dtype<double>) {
       do_rebin_non_inner<double>(dim, var, rebinned, oldCoord, newCoord);
    } else if (oldCoord.dtype() == dtype<float>) {
        do_rebin_non_inner<float>(dim, var, rebinned, oldCoord, newCoord);
    } else {
      throw std::runtime_error(
          "Rebinning is possible only for double and float types.");
    }
```
Given there is only two types at play, it may not be worth changing.

Reply:
I am hoping to eventually replace the `rebin_non_inner` with something using the same kernel (`core::element::rebin`), maybe in combination with `callDType`. Can look into cleaning this up if that succeeds.